include/boost/geometry/algorithms/detail/buffer/implementation.hpp

```diff
@@ -0,0 +1,210 @@
+    <
+        typename DistanceStrategy,
+        typename SideStrategy,
+        typename JoinStrategy,
+        typename EndStrategy,
+        typename PointStrategy,
+        typename Strategies
+    >
+    static inline void apply(Input const& geometry_in,
+                             Output& geometry_out,
+                             DistanceStrategy const& distance_strategy,
+                             SideStrategy const& side_strategy,
+                             JoinStrategy const& join_strategy,
+                             EndStrategy const& end_strategy,
+                             PointStrategy const& point_strategy,
+                             Strategies const& strategies)
+    {
+        typedef typename boost::range_value<Output>::type polygon_type;
+
+        typedef typename point_type<Input>::type point_type;
+        typedef typename rescale_policy_type
+            <
+                point_type,
+                typename geometry::cs_tag<point_type>::type
+            >::type rescale_policy_type;
+
+        if (geometry::is_empty(geometry_in))
+        {
+            // Then output geometry is kept empty as well
+            return;
+        }
+
+        model::box<point_type> box;
+        geometry::envelope(geometry_in, box);
+        geometry::buffer(box, box, distance_strategy.max_distance(join_strategy, end_strategy));
+
+        rescale_policy_type rescale_policy
+                = boost::geometry::get_rescale_policy<rescale_policy_type>(
+                    box, strategies);
+
+        detail::buffer::buffer_inserter<polygon_type>(geometry_in,
+                    range::back_inserter(geometry_out),
+                    distance_strategy,
+                    side_strategy,
+                    join_strategy,
+                    end_strategy,
+                    point_strategy,
+                    strategies,
+                    rescale_policy);
+    }
+};
+
+
+template <typename Input, typename Output>
+struct buffer_all<Input, Output, geometry_collection_tag, multi_polygon_tag>
+{
+    template
+    <
+        typename DistanceStrategy,
+        typename SideStrategy,
+        typename JoinStrategy,
+        typename EndStrategy,
+        typename PointStrategy,
+        typename Strategies
+    >
+    static inline void apply(Input const& geometry_in,
+                             Output& geometry_out,
+                             DistanceStrategy const& distance_strategy,
+                             SideStrategy const& side_strategy,
+                             JoinStrategy const& join_strategy,
+                             EndStrategy const& end_strategy,
+                             PointStrategy const& point_strategy,
+                             Strategies const& strategies)
+    {
+        // NOTE: This algorithm merges partial results iteratively.
+        //   We could first gather all of the results and after that
+        //   use some more optimal method like merge_elements().
```

Comment:
The buffer normally calculates everything at once (by pieces) and traverses all of them to apply the union operation. Not even by merging elements.
But that is complex and has led to issues as well. So I understand that you now do it like this. Maybe forever indeed. OK for me,
but maybe you can include it in the comment.

Reply:
I included your comment.